### PR TITLE
Manage input group state with span onChange instead of element onChange

### PR DESCRIPTION
**Pull request by jlhogan**
the diffs of the group components are hard to look at, but I think the files are pretty readable. May be worth checking out the branch and going thru the code locally.

### CODE REVIEW

File: src/Hedwig/ClientApp/src/components/CheckboxGroup/CheckboxGroup.tsx [hunk 1]
```diff
@@ -93,56 +60,96 @@ export const CheckboxGroup: React.FC<CheckboxGroupProps> = ({
 	horizontal,
 	disabled,
 	...props
-}) => {
+}: CheckboxGroupProps<TFormData>) => {
+	if((typeof ({} as TFormData)) === (typeof ({} as NoFormType))) {
```
Comment (jlhogan) [inline src/Hedwig/ClientApp/src/components/CheckboxGroup/CheckboxGroup.tsx:64]: thoughts on this? it works but i don't know if it's the smelliest thing u ever saw

File: src/Hedwig/ClientApp/src/components/Form_New/FormContext.ts [hunk 2]
```diff
@@ -24,7 +24,7 @@ type FormContextType = {
  */
 export type GenericFormContextType<T> = {
 	data: T;
-	dataDriller: ObjectDriller<T>;
+	dataDriller: TObjectDriller<NonNullable<T>>;
```
Comment (jlhogan) [inline src/Hedwig/ClientApp/src/components/Form_New/FormContext.ts:27]: just tightening up this typing

File: src/Hedwig/ClientApp/src/containers/Enrollment/_sections/ChildInfo/Fields/Race.tsx [hunk 3]
```diff
@@ -44,25 +43,9 @@ const raceOptionFactory: (label: string, field: RaceField) => CheckboxOption = (
 /**
  * Component for entering the race of a child in an enrollment.
  */
-// Each race option is individually stored on the child model. This prevents us
```
Comment (jlhogan) [inline src/Hedwig/ClientApp/src/containers/Enrollment/_sections/ChildInfo/Fields/Race.tsx:47]: IMO this is a totally valid way to use the CheckboxGroup, and I tried to indicate that in comments in that file. I dont know that this needs to be so explained

File: src/Hedwig/ClientApp/src/components/CheckboxGroup/CheckboxGroup.tsx (hunk 1 above)
Comment (jlhogan) [reply src/Hedwig/ClientApp/src/components/CheckboxGroup/CheckboxGroup.tsx:64]: jk it doesnt even work ha ha ha

Comment (cailyncodes) [reply src/Hedwig/ClientApp/src/components/CheckboxGroup/CheckboxGroup.tsx]: Whoa :brain:

File: src/Hedwig/ClientApp/src/components/CheckboxGroup/CheckboxGroup.stories.tsx [hunk 4]
```diff
@@ -1,41 +1,36 @@
 import React from 'react';
 import { storiesOf } from '@storybook/react';
 import { action } from '@storybook/addon-actions';
-import { FormStatusProps } from '..';
+import { FormStatusProps, FieldSet } from '..';
```
Comment (jlhogan) [inline src/Hedwig/ClientApp/src/components/CheckboxGroup/CheckboxGroup.stories.tsx:4]: can go

File: src/Hedwig/ClientApp/src/components/CheckboxGroup/CheckboxGroup.tsx [hunk 5]
```diff
@@ -1,148 +1,133 @@
...
-	);
-};
-
-export type CheckboxGroupProps = InternalCheckboxGroupProps & FieldSetProps;
+export type CheckboxGroupProps<
+	TFieldSetProps extends FieldSetProps | FormFieldSetProps<any>
+> = InternalCheckboxGroupProps &
+	(TFieldSetProps extends FormFieldSetProps<infer T>
+		? { useFormFieldSet: true } & FormFieldSetProps<T>
+		: FieldSetProps);
```
Comment (cailyncodes) [inline src/Hedwig/ClientApp/src/components/CheckboxGroup/CheckboxGroup.tsx:41]: Look at your TypeScript foo!!! :fire:

File: src/Hedwig/ClientApp/src/components/CheckboxGroup/CheckboxGroup.tsx [hunk 6]
```diff
@@ -1,148 +1,133 @@
...
 
 /**
- * Component for displaying a group of related Checkbox items in a FieldSet
+ * Component for displaying a group of related Checkboxes.
+ * Renders the checkbox group inside a FieldSet by default, or FormFieldSet
+ * if
```
Comment (cailyncodes) [inline src/Hedwig/ClientApp/src/components/CheckboxGroup/CheckboxGroup.tsx:46]: Missing end of comment

File: src/Hedwig/ClientApp/src/components/CheckboxGroup/CheckboxGroup.tsx [hunk 7]
```diff
@@ -1,148 +1,133 @@
...
+				{...formFieldSetProps}
+				id={`${id}-fieldset`}
+				childrenGroupClassName={cx(childrenGroupClassName, 'margin-top-3')}
+			>
+				<InternalCheckboxGroup id={id} {...checkboxGroupProps} />
+			</FormFieldSet>
+		);
+	}
+
+	const fieldSetProps = ({ ...props } as unknown) as FieldSetProps;
```
Comment (cailyncodes) [inline src/Hedwig/ClientApp/src/components/CheckboxGroup/CheckboxGroup.tsx:73]: There might be a better way to deal with this union type and unknown casting but I think this is fine

File: src/Hedwig/ClientApp/src/containers/Enrollment/_sections/ChildInfo/Fields/Race.tsx [hunk 8]
```diff
@@ -6,73 +6,21 @@ import { displayValidationStatus } from '../../../../../utils/validations/displa
...
 export const RaceField: React.FC<ChildInfoFormFieldProps> = ({ initialLoad }) => {
-	// We don't provide an onChange prop to CheckboxGroupForForm because the
-	// individual Checkboxes manage their own changes. We also don't include the
-	// onChange prop for each of the options because the FormField component creates
-	// the onChange function and supplies it accordingly.
 	return (
-		<CheckboxGroupForForm<Enrollment>
-			id="race-checkboxgroup"
+		<CheckboxGroup<FormFieldSetProps<Enrollment>>
+			useFormFieldSet
```
Comment (cailyncodes) [inline src/Hedwig/ClientApp/src/containers/Enrollment/_sections/ChildInfo/Fields/Race.tsx:21]: This looks like it might need to be a prop attribute

File: src/Hedwig/ClientApp/src/containers/Enrollment/_sections/FamilyInfo/NewForm.tsx [hunk 9]
```diff
@@ -31,19 +31,7 @@ export const NewForm: React.FC<SectionProps> = ({
...
-			},
-		},
-		// Enrollment needs to be second so it overwrites the empty family if a family exists
-		...enrollment,
-	};
-
-	const [mutatedEnrollment, setMutatedEnrollment] = useState<Enrollment>(
-		initialEnrollment as Enrollment
-	);
+	const [mutatedEnrollment, setMutatedEnrollment] = useState<Enrollment>(enrollment);
```
Comment (cailyncodes) [inline src/Hedwig/ClientApp/src/containers/Enrollment/_sections/FamilyInfo/NewForm.tsx:34]: Why the change here?

File: src/Hedwig/ClientApp/src/components/CheckboxGroup/CheckboxGroup.tsx (hunk 7 above)
Comment (jlhogan) [reply src/Hedwig/ClientApp/src/components/CheckboxGroup/CheckboxGroup.tsx:73]: i'm _sure_ there is but i could not figure it out or google it and i decided since the typescript warning suggests i cast to unknown it is an acceptable solution

File: src/Hedwig/ClientApp/src/containers/Enrollment/_sections/ChildInfo/Fields/Race.tsx (hunk 8 above)
Comment (jlhogan) [reply src/Hedwig/ClientApp/src/containers/Enrollment/_sections/ChildInfo/Fields/Race.tsx:21]: what does that mean?

File: src/Hedwig/ClientApp/src/containers/Enrollment/_sections/FamilyInfo/NewForm.tsx (hunk 9 above)
Comment (jlhogan) [reply src/Hedwig/ClientApp/src/containers/Enrollment/_sections/FamilyInfo/NewForm.tsx:34]: ah yes meant to call this out. the existing spread above does not actually create a non-undefined default family, because spreading the whole enrollment out after will always overwrite with the undefined family value. this was causing and issue when leaving all FamilyInfo fields blank. Made the call to keep the default empty family creation in familyIncome section, since it is required to happen in there (for the case when someone starts creating an enrollment, exits early before family section, then edits familyIncome section view detail view, we need to be able to create a missing family). Here, if the user does not add any family info fields, no family is added, and the app reports that there is "no family information on record" and the enrollment is marked as having missing info

File: src/Hedwig/ClientApp/src/components/CheckboxGroup/CheckboxGroup.tsx (hunk 6 above)
Comment (jlhogan) [reply src/Hedwig/ClientApp/src/components/CheckboxGroup/CheckboxGroup.tsx:46]: oops!
